v5: Floating labels

mdo opened the pull request:
Fixes #21312.

---

- [x] Document more things?
- [x] Confirm browser support
- [x] Look into #27121 (see https://github.com/twbs/bootstrap/pull/30449#issuecomment-704859889)

Preview: <https://deploy-preview-30449--twbs-bootstrap.netlify.com/docs/5.0/forms/floating-labels/>

mdo committed:
v5: Promote floating labels example to component

- Adds new .form-floating
- Stubs out basics of a docs page
- Removes existing Example

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,59 @@
+    &::-webkit-input-placeholder  { color: transparent; }
+    &::-ms-input-placeholder      { color: transparent; }
+    &::-moz-placeholder           { color: transparent; }
+    &::placeholder                { color: transparent; }
```
I think autoprefixer already prefixes these?

XhmikosR replied:
Yup and makes me wonder why Stylelint doesn't error out...

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,59 @@
+    &::placeholder                { color: transparent; }
+
+    &:not(:placeholder-shown) {
+      padding-top: $form-floating-input-padding-t;
```
We shouldn't animate paddings, but use transforms instead

mdo committed:
Update floating labels to support .form-select, make inputs and selects more consistent

- To do this, I made the .form-control and .form-select consistent in min-height vs height
- Removed some unused variables now
- Updated -color to be the -color because I don't know why this was any different before
- Update page to include some examples for layout, validation, and value
- Rewrite styles to not modify padding, but instead transform and opacity

micka-fdz commented:
Good news! Is it possible to add support for `<textarea>` and `.form-file` elements?

XhmikosR commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    display: none;
+  }
+  .form-floating > .form-control::-ms-input-placeholder {
+    color: #777;
```
Is the fallback supposed to be like that here, intact from the example?

mdo replied:
One of the todos I have listed is to check browser support, so I'll know more when I get at this on my PC :).

MartijnCuppens commented:
It feels a bit odd to me that the label doesn't shrink when The input field is focused, it looks like you're going to type in text before the password text:
![image](https://user-images.githubusercontent.com/11559216/77822032-4f690a00-70ef-11ea-9a47-2ee5481ded0a.png)

I would prefer:
![image](https://user-images.githubusercontent.com/11559216/77822059-88a17a00-70ef-11ea-9ca6-bf5bfb4f3f53.png)

Thoughts?

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    cursor: text; // Match the input under the label
+    border: $input-border-width solid transparent;
+    @include border-radius(.25rem);
+    @include transition(all .1s ease-in-out);
```
I would explicitly set `opacity` and `transform` here, like we do with other elements (eg. buttons)

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+  > .form-control:not(:placeholder-shown),
+  > .form-select {
+    ~ label {
+      @include font-size($form-floating-label-font-size);
```
Transitioning the font-size isn't optimal. We better use scale in transform to fix this (we'll probably need to adjust the transform origin too)

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    ~ label {
+      @include font-size($form-floating-label-font-size);
+      opacity: .65;
+      transform: translate3d(0, -.75rem, 0);
```
Using translate3d triggers implicit compositing, which might result in blurry labels. We better use `translateY` here

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    width: 100%;
+    margin-bottom: 0; // Override default `<label>` margin
+    pointer-events: none;
+    cursor: text; // Match the input under the label
```
This is moot since we have `pointer-events: none;`

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    left: 0;
+    display: block;
+    width: 100%;
+    margin-bottom: 0; // Override default `<label>` margin
```
This is moot since we use `position: absolute`

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    position: absolute;
+    top: 0;
+    left: 0;
+    display: block;
```
This is moot since we use `position: absolute`

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    top: 0;
+    left: 0;
+    display: block;
+    width: 100%;
```
Not sure why we need this, but I think we can leave this out too

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    margin-bottom: 0; // Override default `<label>` margin
+    pointer-events: none;
+    cursor: text; // Match the input under the label
+    border: $input-border-width solid transparent;
```
Why do we have this transparent border?

mdo replied:
Added a code comment—it affects the box model and shifts the text.

MartijnCuppens commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,65 @@
+    pointer-events: none;
+    cursor: text; // Match the input under the label
+    border: $input-border-width solid transparent;
+    @include border-radius(.25rem);
```
There is no visible background or border, what's the point of this than?

MartijnCuppens commented on scss/forms/_form-select.scss:
```diff
@@ -6,7 +6,7 @@
   display: block;
   width: 100%;
-  height: $form-select-height;
+  min-height: $form-select-height;
```
We can't do this, because FF adds 2px to the height in a pseudo element we can not control: https://searchfox.org/mozilla-central/rev/c7e8bc4996f979e5876b33afae3de3b1ab4f3ae1/layout/style/res/forms.css#339

alecpl commented:
1. The "placeholder" color is wrong. Should be grey-ish, now it uses input text color. Is this something possible to fix?
2. I think I would prefer the way it looks as proposed in https://github.com/twbs/bootstrap/issues/21312#issuecomment-605595742

mdo committed:
Streamline and bulletproof some things

- Apply some optimizations from code review
- Removed unecessary properties from the label
- Add some comments for what properties are required
- Move from fixed height for labels to height 100% so we can support textareas
- Improve docs a little bit, add ToC

mdo commented:
Pushed some changes based on the code review and fixing some conflicts. Also made the change from @MartijnCuppens's suggestion and moved the label on `:focus` now, too. I dig!

Added textarea support too, @micka-fdz. Unsure about file inputs... those ones are more complicated. Could be an add-on in the future once this gets settled.

mdo committed:
Move some values to variables, switch from scaling font-size to scale, update transforms

mdo commented:
Pushed some more updates to address @MartijnCuppens's comments. Wondering if there's some more I can do to make this more bulletproof. Happy to hear out any comments!

ffoodd commented:
@mdo Regarding browser support and accessibility, I think my PR against our example (#30966) should help:
1. Better fallback for Legacy Edge, avoiding to hide labels,
2. more precise browser support doc.

patrickhlauke commented:
once you/we settled on the definitive version, i can have a quick look accessibility-wise if needed / if you ping me and point me to it.

mdo committed:
Bring over changes from #30966 and add to them to tighten things up

mdo commented:
With the latest changes from @ffoodd's PR, plus some more tweaks, we've got a cleaned up Legacy Edge setup.

<img width="832" alt="Screen Shot 2020-09-23 at 4 34 41 PM" src="https://user-images.githubusercontent.com/98681/94084615-bde70600-fdba-11ea-8c89-5d54510f4c54.png">

XhmikosR committed:
Delete the now unused example images

XhmikosR commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,70 @@
+  .form-floating > label {
+    position: static;
+    padding: 0;
+    margin-bottom: calc(#{$form-floating-padding-y} / 2);
```
Are we supposed to use `calc` here or our function?

XhmikosR commented on scss/_variables.scss:
```diff
@@ -640,7 +640,7 @@ $input-bg:                              $white !default;
 $input-disabled-border-color:           null !default;
 
-$input-color:                           $gray-700 !default;
+$input-color:                           $body-color !default;
```
Just to confirm that this change is intended here?

mdo replied:
Yup, decided there's no good reason to lighten the color or detach it from the body.

XhmikosR commented on site/content/docs/5.0/forms/overview.md:
```diff
@@ -18,6 +18,8 @@ sections:
   - title: Input group
     description: Attach labels and buttons to your inputs for increased semantic value.
+  - title: Floating labels
+    description: Create beautifully simple forms labels that float over your input fields.
```
Should "forms" be "form" here and in the new page description?

patrickhlauke replied:
Yes it should be singular

XhmikosR committed:
Fix typo

mmrath commented:
I was looking at the examples from deployed site. I see that when the input is empty, placeholder is not shown. It just shows the label. It would be nice if the placeholder text can be shown with the label always floating on top, something like the screenshot below.

![image](https://user-images.githubusercontent.com/115767/95033967-bce39c00-070b-11eb-83ee-6c8b9a11e8e9.png)

patrickhlauke commented:
@mmrath mixing both placeholder and floating label makes no sense, unless i'm misunderstanding what you're after... by design, floating labels initially appear visually as if they were in the place of the form input/placeholder text, and then move out of the way when the user wants to enter something.

sounds more like you're after a style for labels that makes them appear small/as part of the input, statically?

mmrath commented:
@patrickhlauke Thanks for your reply. I could be wrong or misunderstanding the feature but I'll try to explain what I am thinking.

With the code below I was expecting that `name@example.com` would appear as a placeholder and `Email address` as label.  

```
<div class="form-floating mb-3">
  <input type="email" class="form-control" id="floatingInput" placeholder="name@example.com">
  <label for="floatingInput">Email address</label>
</div>
```

patrickhlauke commented:
floating labels are labels that initially appear like an entry, but when clicked move to the side/up/become smaller 

![floating labels animation](https://miro.medium.com/max/500/1*GqNGB0mTjPwsRXInOlCZUw.gif)

(from https://medium.com/simple-human/floating-labels-are-a-bad-idea-82edb64220f6)

patrickhlauke commented:
but now looking at even the first example on https://deploy-preview-30449--twbs-bootstrap.netlify.app/docs/5.0/forms/floating-labels/ with the "Email address" field, do you mean you'd expect the placeholder to show once the user clicks into the input, the floating label itself moves out of the way, but the user hasn't entered anything yet? in which case yeah, looking just at the code there that does indeed include a `placeholder`, that would probably be expected behavior - not sure if we're explicitly suppressing this for a reason of not...

mmrath commented:
I did a bit more digging, I think I was expecting it to work like "Infield Top-Aligned Labels" as per the article below. So as per the article current "floating labels" are probably implemented correctly in this PR. 
As you said currently the placeholder is totally unused (except for `:placeholder-shown` usage). I personally think 'Infield Top-Aligned Labels" would have been better, but I am not a UX person.

https://uxmovement.com/forms/infield-top-aligned-labels-floating-labels/

screenshot from the website
![image](https://user-images.githubusercontent.com/115767/95078051-40c77380-0760-11eb-8a7d-f6364f843402.png)

MartijnCuppens commented:
#27121 seems to be fixed:
![image](https://user-images.githubusercontent.com/11559216/95322473-d536f080-089c-11eb-8e23-07d6912a5e89.png)

mmrath commented:
@MartijnCuppens #27121 appears to be related to value of the input field not placeholder.

HamedFathi commented:
Guys, Is it possible to have a smaller version of floating inputs? I think the final result is so big!

I think [this way](https://pryley.github.io/float-labels.js) was so flexible to use any size of inputs. (sm, lg, default)

```html
<input class="form-control form-control-lg" type="text">
<input class="form-control" type="text">
<input class="form-control form-control-sm">
```

![image](https://user-images.githubusercontent.com/8418700/96331676-04531c00-106c-11eb-8dec-b1425e964398.png)

mdo committed:
Allowlist the calc function

mdo committed:
Add transform-origin, update transform values

eduardogoncalves commented:
@mdo It's working now
![image](https://user-images.githubusercontent.com/4308648/97234145-9c60ac00-17b6-11eb-9ae7-703e44b62822.png)

mdo committed:
Test out autofill fix

mdo committed:
Fix linter issue

mdo commented:
@eduardogoncalves Yup, pushed a new fix to override that last commit with separated rules. Should be good now.

mdo committed:
Mention it in the migration guide

mdo commented:
This should be good to go now! Can I get a final review?

mdo committed:
Bump bundlesize

alecpl commented:
I think I don't like these big inputs (and with strange padding when they are empty). Something like in https://github.com/twbs/bootstrap/pull/30449#issuecomment-710769866 would be better.

micka-fdz commented:
I agree, fields are currently much too big.
I prefer this render : https://getbootstrap.com/docs/4.5/examples/floating-labels/

MartijnCuppens commented:
https://github.com/twbs/bootstrap/pull/30449#issuecomment-710769866 would require an additional background to fix the border and shadow overlap

![image](https://user-images.githubusercontent.com/11559216/97283649-636d1980-1840-11eb-9739-6db9796aeec3.png)

This will look weird if the parent element has another background
![image](https://user-images.githubusercontent.com/11559216/97283858-a62ef180-1840-11eb-80fb-331b567a97c0.png)

Or if the background of the input is semi-transparent:
![image](https://user-images.githubusercontent.com/11559216/97284278-25bcc080-1841-11eb-84ea-0b636e800c02.png)


But I can agree the input looks a little to big. https://getbootstrap.com/docs/4.5/examples/floating-labels/ might indeed be a better look as @micka-fdz suggests.

alecpl commented:
Semi-transparent input is a rare-case that could be just unsupported by floating labels. The other label background issue I could live with if the padding-top was smaller (and maybe with rounded corners). Still I would prefer this over what's proposed in the PR.

alecpl commented:
@mdo, in firefox 82 on linux the label animation has some issue. When you focus the input there are two steps of the animation visible, first the label moves up, then it's font size drops to even smaller. Maybe it's a firefox's transform/scale bug. Looks like when the text is moving it is blured then it sharpens. I don't see this in Chrome.

ffoodd commented on scss/forms/_floating-labels.scss:
```diff
@@ -0,0 +1,85 @@
+    pointer-events: none;
+    border: $input-border-width solid transparent; // Required for aligning label's text with the input as it affects inner box model
+    transform-origin: 0 0;
+    @include transition(opacity .1s ease-in-out, transform .1s ease-in-out);
```
Shouldn't this value be variable, as we do with `$input-transition`?
Apart from that, seems fine!

ffoodd replied:
Trying this on both Firefox & Chromium on Ubuntu, I can see a little font rendering jump (more obvious on Firefox). I guess it can be related to the font-stack I updated recently—already mentionned a related bug in #31955 by @MartijnCuppens.

That's unrelated to the PR I guess, I'd need more testing on those using our old font-stack and compare other possibilities… Opening an issue to track this.

mdo committed:
Add one more variable per review

mdo committed:
Shave .25rem off the height

micka-fdz commented:
Hi @MartijnCuppen, what about my suggestion of inputs height? I think it’s a quick modification. Maybe can I help?
